primary/public-src/ts/script-editform.ts

```diff
@@ -0,0 +1,229 @@
+					}
+					for (let varName in op.quantifiers) {
+						lightAssert(thisMultiselectOpts.includes(varName), `Operation #${i}, multiselect, has a quantifier for the string ${varName}, but that string is not an option for the multiselect ${derivedDescription}`);
+						// Also, check that the quantifier is a number
+						lightAssert(typeof op.quantifiers[varName] === 'number', `Operation #${i}, multiselect, has a non-number quantifier "${op.quantifiers[varName]}" for the string ${varName} ${derivedDescription}`);
+					}
+					break;
+				}
+				// Operations that take 2 operands
+				case 'subtract': case 'divide': case 'gt': case 'gte': case 'lt': case 'lte': case 'eq': case 'ne': case 'min': case 'max': {
+					let op = thisOp as | DivideOperation | CompareOperation | MinMaxOperation;
+					lightAssert(op.operands.length === 2, `Operation #${i}, ${op.operator}, needs exactly 2 operands ${derivedDescription}`);
+					validateOperands(op.operands);
+					break;
+				}
+				// Operations that take N operands
+				case 'add': case 'sum': case 'multiply': {
+					let op = thisOp as SumOperation | MultiplyOperation;
+					validateOperands(op.operands);
+					break;
+				}
+				// 1 operand
+				case 'abs': {
+					let op = thisOp as AbsoluteValueOperation;
+					lightAssert(op.operands.length === 1, `Operation #${i}, ${op.operator}, needs exactly 1 operand ${derivedDescription}`);
+					validateOperands(op.operands);
+					break;
+				}
+				// 1 operand then 1 number
+				case 'log': {
+					let op = thisOp as LogOperation;
+					lightAssert(op.operands.length === 2, `Operation #${i}, ${op.operator}, needs exactly 2 operands ${derivedDescription}`);
+					validateOperands([op.operands[0]]);
+					lightAssert(typeof op.operands[1] === 'number', `Operation #${i}, ${op.operator}, needs its second operand to be a number but found "${op.operands[1]}" ${derivedDescription}`);
+				}
+			}
+			// Add the output of this operation to the list of intermediate variables
+			if (i < derived.operations.length - 1) {
+				let thisOpAs = thisOp.as;
+				lightAssert(thisOpAs, `Operation #${i}, ${thisOp.operator} does not have an 'as' keyword ${derivedDescription}`);
+				intermediateVariables.push('$' + thisOpAs);
+			}
+		});
+	}
+
+	// verify keys are all 'allowed'
+	// TODO
+
+
+	// check if any existing data fields are not included
+	// TODO enable for Pit Scouting (will need changes in orgconfig.ts as well)
+	let missingDataKeys = previousKeys.filter((key) => !idDict[key]);
+	if (missingDataKeys.length > 0) {
+		let message = `*WARNING!* The following keys (ids) exist in your past match scouting data, but are not included in this layout: \n${missingDataKeys.map(key => `\n - ${key}`)}`;
+		let { cancelled } = await Confirm.show(message, { yesText: 'OK', noText: 'Cancel' });
+		if (cancelled) return null;
+	}
+	return jsonString;
+}
+
+async function test() {
+	let jsonString = await validate();
+	if (jsonString == null)
+		return;
```

Comment:
does `lightAssert` raise an error if it fails? if so, where is it handled? does it also make the function return null, or is it skippable?

Reply:
Yes, it does throw; The difference between lightAssert and assert are that assert shows a stack trace and a "Please report as an issue on our GitHub page", while lightAssert just shows the message text. The idea is that assert is to be used when we **don't** expect it to fail, i.e. failing the assert is some unexpected error; while lightAssert is to be used when we can reasonably expect the error to occur sometimes (like in the case of the user providing bad input). May be a good idea to rename it to make that more clear. Any ideas?

Reply:
Yeah that part makes sense, I was more asking about the code part where lightAssert has the throw keyword but there doesn't seem to be any try block to catch it so what effect does that have on the main UI code or anything below it. It might be fine cuz this is more vanilla but in something like Svelte it could cause issues.

Reply:
I think that in the case of our vanilla JS/TS environment, the only thing that'll happen aside from the NotificationCard being shown is that the unhandled error will be logged to the console.